### PR TITLE
Fix: RL base feature parity

### PR DESCRIPTION
**Pull request by NanoCode012**
# Description

RL trainer was not loading some basic configs like `logging_steps` etc. This PR consolidates the setting of these params and cleans them up.

This PR also fixes a case where we did not call `.map` with `num_proc`.

To discuss:
- [x] Handling of bf16/bfloat16
- [x] Handling of fp16 in RL
- [x] Handling of tf32 in RL
- [x] Default of `use_reentrant`
- [x] Default of `remove_unused_columns`

Addresses a lot of points in #2105 

## How has this been tested?

~Untested!~

## Summary by CodeRabbit

- **New Features**
  - Added new trainer builder modules for both causal language modeling and reinforcement learning workflows, offering more modular and extensible configuration options.
  - Introduced support for a new configuration parameter `torch_compile_mode` and an optional field `dpo_label_smoothing` for advanced training customization.

- **Refactor**
  - Restructured trainer builder logic into dedicated modules, improving maintainability and clarity.
  - Updated import paths throughout the codebase to reflect the new modular trainer builder structure.
  - Enhanced optimizer integration for trainers using mixins for more flexible initialization and configuration.

- **Bug Fixes**
  - Corrected reward function handling and argument blocklisting in GRPO strategy to improve reliability.

- **Documentation**
  - Updated documentation configuration to reflect the new builder module structure and clarified certain configuration options.

- **Tests**
  - Added comprehensive unit and integration tests for the new trainer builder modules, covering various training configurations and optimizer scenarios.

### CODE REVIEW

**Review by bursteratom — APPROVED** [review]
LGTM

**Review by winglian — APPROVED** [review]

**Comment by winglian** [comment]
Thanks @NanoCode012 ! This should be good to go once the multi gpu tests pass too https://github.com/axolotl-ai-cloud/axolotl/actions/runs/13457763772

**Comment by winglian** [comment]
latest ci: https://github.com/axolotl-ai-cloud/axolotl/actions/runs/13469023440

**Comment by winglian** [comment]
looks like the multi-gpu GRPO tests are failing https://github.com/axolotl-ai-cloud/axolotl/actions/runs/13469023440/job/37640091555

File: src/axolotl/core/trainer_builder.py [hunk 1]
```diff
@@ -224,6 +224,155 @@ def hook_post_create_trainer(self, trainer):
...
+            warmup_steps = max(int(self.cfg.warmup_ratio * total_num_steps), 0)
+        else:
+            warmup_steps = min(int(0.03 * total_num_steps), 100)
+        if warmup_steps == 1:
+            warmup_steps = 2
+
+        logging_steps = (
+            self.cfg.logging_steps
+            if self.cfg.logging_steps is not None
+            else max(min(int(0.005 * total_num_steps), 10), 1)
```
Comment (winglian) [inline src/axolotl/core/trainer_builder.py]: for many of the RL trainers, we can't rely on total_num_steps (set to None) and we pass -1 as max_steps to let the trainer figure out the total number of steps

File: src/axolotl/core/trainer_builder.py [hunk 2]
```diff
@@ -224,6 +224,155 @@ def hook_post_create_trainer(self, trainer):
...
+
+        warmup_steps = None
+        if self.cfg.warmup_steps is not None:
+            warmup_steps = self.cfg.warmup_steps
+        elif self.cfg.warmup_ratio is not None:
+            warmup_steps = max(int(self.cfg.warmup_ratio * total_num_steps), 0)
+        else:
+            warmup_steps = min(int(0.03 * total_num_steps), 100)
+        if warmup_steps == 1:
+            warmup_steps = 2
```
Comment (winglian) [inline src/axolotl/core/trainer_builder.py]: what is this for?

**Comment by winglian** [comment]
https://github.com/axolotl-ai-cloud/axolotl/actions/runs/13485689047

File: src/axolotl/core/trainer_builder.py (hunk 2 above)
Comment (NanoCode012) [reply src/axolotl/core/trainer_builder.py]: This was in the original SFT trainer https://github.com/axolotl-ai-cloud/axolotl/pull/2133/files#diff-5edc13801ecfdd108e81872527bdc78c6d24a73833968147b0a9ecb8452996f4L323-L324

**Comment by djsaunde** [comment]
One suggestion: can we break out the `trainer_builder.py` file into smaller modules? E.g., something like:

```shell
builders
├── __init__.py
├── base.py
├── causal.py
└── rl.py
```

**Comment by NanoCode012** [comment]
> One suggestion: can we break out the `trainer_builder.py` file into smaller modules? E.g., something like:
> 
> ```shell
> builders
> ├── __init__.py
> ├── base.py
> ├── causal.py
> └── rl.py
> ```

Yep, I can do that tomorrow!

**Comment by NanoCode012** [comment]
Since the PR is quite old, rebasing proved to be a bit challenging, so I'm swapping to merge from main.

I think this PR is ready to be re-reviewed / almost good to go.

TODO left:
- [ ] whether to set `eval_accumulation_steps` to grad accu in RL (like in SFT)

File: _quarto.yml [hunk 3]
```diff
@@ -17,7 +17,9 @@ quartodoc:
         - convert
         - prompt_tokenizers
         - logging_config
-        - core.trainer_builder
```
Comment (djsaunde) [inline _quarto.yml:20]: I kind of prefer `core.builders.base`, etc. naming, what do you think?

File: docs/config.qmd [hunk 4]
```diff
@@ -529,7 +530,7 @@ profiler_steps: # enable the pytorch profiler to capture the first N steps of tr
 loss_watchdog_threshold: # High loss value, indicating the learning has broken down (a good estimate is ~2 times the loss at the start of training)
 loss_watchdog_patience: # Number of high-loss steps in a row before the trainer aborts (default: 3)
 
-# Save model as safetensors (require safetensors package)
```
Comment (djsaunde) [inline docs/config.qmd:563]: What does "Transformers default True" mean?

File: src/axolotl/core/trainer_builder/__init__.py
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/__init__.py:1]: nit: can you import the classes you intend to expose here and put them in `__all__`? Then, you can shorten the import statement elsewhere:

```python
from axolotl.core.trainer_builder import ...
```

File: src/axolotl/core/trainer_builder/base.py [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+
+    def _configure_precision_settings(self, training_args_kwargs):
+        training_args_kwargs["fp16"] = (self.cfg.fp16 and not self.cfg.bf16) or False
+        training_args_kwargs["tf32"] = self.cfg.tf32
+        if self.cfg.bf16 == "full":
+            training_args_kwargs["bf16_full_eval"] = True
+        else:
+            training_args_kwargs["bf16"] = self.cfg.bf16 or self.cfg.bfloat16
+
+    def _configure_optimizer_and_scheduler(self, training_args_kwargs):
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/base.py]: IMO you could break this out into two methods, this is pretty large.

File: src/axolotl/core/trainer_builder/base.py [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+        if self.cfg.optimizer in custom_supported_optimizers:
+            # Common optimizer kwargs
+            optimizer_kwargs = {
+                "lr": training_args_kwargs["learning_rate"],
+                "weight_decay": training_args_kwargs["weight_decay"],
+            }
+
+            # Adam-specific kwargs
+            adam_kwargs: dict = {}
+            if training_args_kwargs.get("adam_beta1") and training_args_kwargs.get(
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/base.py]: why can't the user specify one of these params and not the other?

File: src/axolotl/core/trainer_builder/base.py [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+    def _set_base_training_args(self, total_num_steps) -> dict[str, Any]:
+        training_args_kwargs: Dict[str, Any] = {}
+
+        self._configure_warmup_and_logging(total_num_steps, training_args_kwargs)
+
+        self._configure_precision_settings(training_args_kwargs)
+
+        self._configure_save_and_eval_strategy(training_args_kwargs)
+
+        self._configure_gradient_checkpointing(training_args_kwargs)
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/base.py]: personally I prefer:

```suggestion
        self._configure_warmup_and_logging(total_num_steps, training_args_kwargs)
        self._configure_precision_settings(training_args_kwargs)
        self._configure_save_and_eval_strategy(training_args_kwargs)
        self._configure_gradient_checkpointing(training_args_kwargs)
```

File: src/axolotl/core/trainer_builder/base.py [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+
+        # set arg into trainer_args_kwargs with same name if value not None
+        for arg in [
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/base.py]: IMO this should be a constant

File: src/axolotl/core/trainer_builder/base.py [hunk 9]
```diff
@@ -0,0 +1,496 @@
...
+        if self.cfg.reward_model or self.cfg.rl:
+            training_args_kwargs["max_length"] = self.cfg.sequence_len
+
+        self._configure_reporting(training_args_kwargs)
+
+        self._configure_hub_parameters(training_args_kwargs)
+
+        self._configure_optimizer_and_scheduler(training_args_kwargs)
+
+        self._configure_torch_compile(training_args_kwargs)
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/base.py]:
```suggestion
        self._configure_reporting(training_args_kwargs)
        self._configure_hub_parameters(training_args_kwargs)
        self._configure_optimizer_and_scheduler(training_args_kwargs)
        self._configure_torch_compile(training_args_kwargs)
```

File: src/axolotl/core/trainer_builder/base.py [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+        training_args_kwargs["per_device_train_batch_size"] = self.cfg.micro_batch_size
+
+        if self.cfg.eval_batch_size:
+            training_args_kwargs["per_device_eval_batch_size"] = (
+                self.cfg.eval_batch_size
+            )
+
+        training_args_kwargs["max_steps"] = self.cfg.max_steps or total_num_steps or -1
+
+        training_args_kwargs["num_train_epochs"] = self.cfg.num_epochs
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/base.py]:
```suggestion
        training_args_kwargs["max_steps"] = self.cfg.max_steps or total_num_steps or -1
        training_args_kwargs["num_train_epochs"] = self.cfg.num_epochs
```

File: src/axolotl/core/trainer_builder/rl.py [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+            trainer_cls_args = [self.model]
+        elif self.cfg.rl is RLType.SIMPO:
+            trainer_cls = AxolotlCPOTrainer
+            trainer_cls_args = [self.model]
+        else:
+            raise ValueError(f"Unsupported RL: {self.cfg.rl}")
+
+        return trainer_cls, trainer_cls_args
+
+    def build_training_arguments(self, total_num_steps):
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/rl.py]: Could be private

File: src/axolotl/core/trainer_builder/rl.py [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+        training_args = training_args_cls(  # pylint: disable=unexpected-keyword-arg
+            logging_first_step=True,
+            **training_args_kwargs,
+        )
+
+        # unset run_name so wandb sets up experiment names
+        if self.cfg.use_wandb and training_args.run_name == training_args.output_dir:
+            training_args.run_name = (  # pylint: disable=attribute-defined-outside-init
+                None
+            )
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/rl.py]: What does this do exactly?

File: src/axolotl/core/trainer_builder/rl.py [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+                ensure_dtype(trainer.ref_model, dtype=self.cfg.torch_dtype)
+
+        trainer = self.hook_post_create_trainer(trainer)
+        for callback in self.get_post_trainer_create_callbacks(trainer):
+            trainer.add_callback(callback)
+
+        return trainer
+
+
+class HFPPOTrainerBuilder(TrainerBuilderBase):
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/rl.py]: Don't think this is used anywhere

File: src/axolotl/core/trainer_builder/sft.py [hunk 14]
```diff
@@ -0,0 +1,490 @@
...
+LOG = logging.getLogger(__name__)
+
+
+class HFCausalTrainerBuilder(TrainerBuilderBase):
+    """
+    Build the HuggingFace training args/trainer for causal models and reward modeling
+    using TRL.
+    """
+
+    def get_callbacks(self):
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/sft.py]: can be private

File: src/axolotl/core/trainer_builder/sft.py [hunk 15]
```diff
@@ -0,0 +1,490 @@
...
+            and self.model.use_bettertransformer is True
+        ):
+            callbacks.append(SaveBetterTransformerModelCallback())
+
+        if self.cfg.loss_watchdog_threshold is not None:
+            callbacks.append(LossWatchDogCallback(self.cfg))
+
+        return callbacks
+
+    def get_post_trainer_create_callbacks(self, trainer):
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/sft.py]: can be private

File: src/axolotl/core/trainer_builder/sft.py [hunk 16]
```diff
@@ -0,0 +1,490 @@
...
+            trainer.add_callback(callback)
+
+        if self.cfg.deepspeed and self.cfg.sample_packing:
+            trainer.accelerator.state.deepspeed_plugin.deepspeed_config[
+                "train_micro_batch_size_per_gpu"
+            ] = self.cfg.micro_batch_size
+
+        return trainer
+
+    def build_collator(
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder/sft.py]: could be private

File: src/axolotl/core/trainers/grpo/__init__.py [hunk 17]
```diff
@@ -167,4 +167,4 @@ def get_reward_func(cls, reward_func_fqn: str) -> RewardFunc:
             LOG.info(
                 f"Reward function {reward_func_fqn} is a pre-trained model path - if this is unexpected, please check the reward function path."
             )
-            return reward_func
+            return reward_func_fqn
```
Comment (djsaunde) [inline src/axolotl/core/trainers/grpo/__init__.py:170]: fqn -> fn?

**Review by djsaunde — COMMENTED** [review]
Great stuff!

Mostly low impact comments from me.

File: src/axolotl/core/trainer_builder/rl.py [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+        """
+        if self.cfg.plugins:
+            plugin_manager = PluginManager.get_instance()
+            trainer_cls = plugin_manager.get_trainer_cls(self.cfg)
+            trainer_cls_args = []  # type: ignore
+
+            if trainer_cls is not None:
+                return trainer_cls, trainer_cls_args
+
+        if self.cfg.rl is RLType.GRPO:
```
Comment (salmanmohammadi) [inline src/axolotl/core/trainer_builder/rl.py]: nit: set `trainer_cls_args = [self.model]` once and extend/append as appropriate

File: _quarto.yml (hunk 3 above)
Comment (NanoCode012) [reply _quarto.yml:20]: I'm open to this. Would we need to adjust the class to be named BuilderBase from TrainerBuilderBase? It may be a bit ambiguous what Builder refers to.

File: docs/config.qmd (hunk 4 above)
Comment (NanoCode012) [reply docs/config.qmd:563]: It means the default upstream is true. I'll update it to be clearer

File: src/axolotl/core/trainer_builder/__init__.py
Comment (NanoCode012) [reply src/axolotl/core/trainer_builder/__init__.py:1]: Yup, will make this change!

File: src/axolotl/core/trainer_builder/base.py (hunk 5 above)
Comment (NanoCode012) [reply src/axolotl/core/trainer_builder/base.py]: I split scheduler out, but not sure if that's enough

File: src/axolotl/core/trainer_builder/base.py (hunk 6 above)
Comment (NanoCode012) [reply src/axolotl/core/trainer_builder/base.py]: Hm, `betas` require a list of beta1 and beta2. We can certainly check for beta1, but if beta2 exist but not beta1, I'm not sure if we should have a default for this.

File: src/axolotl/core/trainer_builder/rl.py (hunk 13 above)
Comment (NanoCode012) [reply src/axolotl/core/trainer_builder/rl.py]: We also don't use `TRLPPOTrainer` but I'll leave this for a separate PR to discuss remove these.

File: src/axolotl/core/trainer_builder/rl.py (hunk 12 above)
Comment (NanoCode012) [reply src/axolotl/core/trainer_builder/rl.py]: I think, if you run experiment with RLHF previously to wandb, it would set the run name to `outputs/lora` etc . This would unset the `run_name` so that wandb would random the name `{random_animal_name}`.

I think this may be because Trainer internally sets `run_name={output_dir}` if not provided.

File: src/axolotl/core/trainer_builder/base.py (hunk 8 above)
Comment (NanoCode012) [reply src/axolotl/core/trainer_builder/base.py]: I don't think it would be reused anywhere else, so would a constant be useful?

File: src/axolotl/core/trainer_builder/sft.py (hunk 14 above)
Comment (NanoCode012) [reply src/axolotl/core/trainer_builder/sft.py]: If the child class calls, `super()._get_callbacks()` , would this make linter complain accessing private attribute? I was mid-refactoring the other functions when I got the above thought.

File: src/axolotl/core/trainers/grpo/__init__.py (hunk 17 above)
Comment (NanoCode012) [reply src/axolotl/core/trainers/grpo/__init__.py:170]: This is intended. If the attempt to load the `reward_func` failed, then we assume `reward_func_fqn` is the valid path

**Comment by NanoCode012** [comment]
In https://github.com/axolotl-ai-cloud/axolotl/pull/2133/commits/e07c33c5603afa1fade6d49b539ea93918c0e784 and https://github.com/axolotl-ai-cloud/axolotl/pull/2133/commits/0e52b4d68bfadd76484701b0f2dab351bc15a52b , I have added parity for setting custom optimizer in RL and RM/PRM.

I think I addressed most of the comments left as well with one under discussion for rename from `trainer_builder->builders` and refactoring to private functions within the builders.

File: _quarto.yml (hunk 3 above)
Comment (djsaunde) [reply _quarto.yml:20]: IMO you should keep the naming as is

**Review by djsaunde — APPROVED** [review]
IMO good to go once the aforementioned naming change is in. Great stuff!

**Comment by NanoCode012** [comment]
Thanks for all the reviews. It was big PR. If there are any missing gaps, I'll address them in a follow up PR.
